multicluster/controllers/multicluster/clusterset_conversion_controller.go

```diff
@@ -0,0 +1,186 @@
+
+import (
+	"context"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"antrea.io/antrea/multicluster/apis/multicluster/v1alpha2"
+	mcsv1alpha2 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha2"
+)
+
+// In Antrea 1.13, ClusterClaim CRD will be removed from Antrea Multi-cluster API.
+// ClusterSetConversionController is responsible to convert fetch clusterID and clusterSetID
+// from existing ClusterClaims into the clusterID field of ClusterSet for smooth upgrade
+// without manual intervention.
+type ClusterSetConversionController struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	namespace string
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+	// existingClusterSet saves the existing ClusterSet in case it's deleted successfully but
+	// the new one is not created yet due to creation error. ClusterSetConversionController
+	// should retry with this existingClusterSet.
+	existingClusterSet *v1alpha2.ClusterSet
+}
+
+func NewClusterSetConversionController(
+	Client client.Client,
+	Scheme *runtime.Scheme,
+	namespace string,
+) *ClusterSetConversionController {
+	controller := &ClusterSetConversionController{
+		Client:    Client,
+		Scheme:    Scheme,
+		namespace: namespace,
+		queue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ClusterSetConversionController"),
+	}
+	return controller
+}
+
+// Enqueue will be called after ClusterSetConversionController is initialized.
+func (c *ClusterSetConversionController) Enqueue() {
+	// The key can be anything as we only have single item.
+	c.queue.Add("key")
+}
+
+// Run starts the ClusterSetConversionController and blocks until stopCh is closed.
+// it will run only once if no error happens.
+func (c *ClusterSetConversionController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting ClusterSetConversionController")
+	defer klog.InfoS("Shutting down ClusterSetConversionController")
+
+	if err := c.RunOnce(); err != nil {
+		c.Enqueue()
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *ClusterSetConversionController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ClusterSetConversionController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.conversion()
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	klog.ErrorS(err, "Error converting ClusterSet, re-queuing it")
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *ClusterSetConversionController) RunOnce() error {
+	err := c.conversion()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *ClusterSetConversionController) conversion() error {
```

Comment:
Could we check the ClusterSet resource API version, create a new v1alpha2 ClusterSet and ignore the existing v1alpha1 ClusterSet? I would avoid modifying or deleting user created v1alpha1 ClusterSets.

Reply:
I am afraid auto conversion won't work without modification especially when the existing ClusterSet CR's name matches the ClusterSet ID. When the name matches the ClusterSet ID, either we delete existing one and create a new one (because of name conflicts), or we need to modify the existing CR with new ClusterID field.
And we have a ClusterSet webhook to allow only one ClusterSet in one namespace. So when the name doesn't match the ClusterSet ID, we need to delete the existing one before create a new one.

Note: We won't be able to ignore the existing v1alpha1 CRs since not matter the user run `kubectl get clusterset` or use go-client to get ClusterSet list, it will return both v1alpha1 and v1alpa2 CRs because we set both version's `served` as `true` and conversion strategy as `None`, K8s api server will convert them automatically. User will always see a full list which combines two versions.
If we set the v1alpha1's `served` as false, it will be impossible to get those existing v1alpha1 CRs, then there is no chance to do ClusterSet conversion since we need leader access info from existing one.

Reply:
I guess the problem is when user already created a ClusterSet CR using ClusterSet ID as the name? If the name is different, we can always create a new ClusterSet right? Are we able to get the APIGroup version for a resource with the go-client?

Reply:
I haven't found a reliable way to tell what's the original APIGroup version for a resource yet. There are client-side annotation and server-side annotation which should include the original version like below, but I doubt it's not always true to have these annotations, trying to check with @tnqn to get more ideas.
```
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"multicluster.crd.antrea.io/v1alpha1","kind":"ClusterSet","metadata":{"annotations":{},"name":"test-clusterset-1","namespace":"kube-system"},"spec":{"leaders":[{"clusterID":"test-cluster-leader","secret":"leader-access-token","server":"https://172.18.0.3:6443"}],"namespace":"antrea-multicluster"}}
  creationTimestamp: "2023-07-11T01:25:39Z"
  generation: 1
  managedFields:
  - apiVersion: multicluster.crd.antrea.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
      f:spec:
        .: {}
        f:leaders: {}
        f:namespace: {}
    manager: kubectl-client-side-apply
    operation: Update
    time: "2023-07-11T01:25:39Z"
```

Reply:
Ok. @tnqn : we updated a CRD from v1alpha1 to v1alpha2 and Lan's PR tries to do an one-time conversion of CRs from old to new version at controller startup. Could you also suggest the right way to handle such CRD version changes?

Reply:
Hi @jianjuns  I have synced with @tnqn , he suggested to provide an antctl sub-command to do this kind of upgrade, I feel it make sense to involve user to do this kind of CR upgrade manually via a one-line command since those CRs are created by user.
I will submit a new PR to add a new command `antctl mc upgrade clusterset`, let me know if you have any concern. Thanks.

Reply:
Does this have the same issue as the earlier manual steps - it needs to be done before deploying the new version of mc-controller? Maybe it is ok, if conversion to new version will not impact old version mc-controller functioning?

How if after the conversion, we restart the old version mc-controller? What will happen in that case?

Reply:
The old version of ClusterSet is still served, so when the old mc-controller is running, it actually still see the old version of ClusterSet as before (The new field ClusterID will be stripped by K8s), the difference is that the back-end storage version is v1apha2 after any CR modification. 
I was thinking to do the ClusterSet upgrade after the new manifest deployment is done. When the new mc-controller is running, it will return error due to no ClusterID in the existing ClusterSet, it will retry until we finished the ClusterSet CR upgrade. It will be back to normal after that.
Another way is to split member manifest into two parts like leader manifest, one is CRDs, another is deployment related resources, then we can guide user to deploy new CRDs first, then upgrade ClusterSet CR, and do deployment upgrade finally.

Reply:
I think we discussed this before - if we deploy new mc-controller first, will it be able to resume functioning after the CR conversion (by antctl), without missing any change in the ClusterSet, including changes in the local cluster or remote clusters? And stale resources if any will be cleaned up too?

Before the conversion, changes in the ClusterSet will be ignored? But existing Multi-cluster Pod/Service traffic can still be going?

If the answers are all true, we can consider the approach. Could we add some logic to handle it more elegantly, e.g. generating logs like waiting for ClusterSet CR conversion, rather than an invalid CR error? Again, it would be better if mc-controller can get the CR version to know it is indeed old version CR.

Reply:
Could you answer my question and explain what downtime we can have?

Reply:
Regarding your previous questions:
1. ` if we deploy new mc-controller first, will it be able to resume functioning after the CR conversion (by antctl), without missing any change in the ClusterSet, including changes in the local cluster or remote clusters? And stale resources if any will be cleaned up too?`

The downtime with original design is ClusterSet controller won't set up a new common area until a new valid ClusterSet CR is ready. The controller can resume functioning after CR conversion but it may miss some changes if the downtime is long. The local change won't be reflected to leader and leader's change won't be imported in local. In this case, stale resource won't be cleaned up until controller is restarted since the stale controller is one-time process if it's succeed at pod start.

2. `Before the conversion, changes in the ClusterSet will be ignored? But existing Multi-cluster Pod/Service traffic can still be going?`
Yes, the change in ClusterSet will be ignored. The ClusterSet controller will watch and handle ClusterSet change, but in current implementation in main branch, if there is no valid ID in the spec, it will ignore the ClusterSet and return error directly. 

I will refine the ClusterSet controller to check the ClusterID from a ClusterClaim, so it can handle the ClusterSet with empty ClusterID. In this way, event we upgrade both CRDs and deployment in one time, the ClusterSet can work properly with old ClusterSet.